proposals/mpi-operator-proposal.md

```diff
@@ -0,0 +1,224 @@
+## Motivation
+Kubeflow currently supports distributed training of TensorFlow models using
+[tf-operator](https://github.com/kubeflow/tf-operator), which relies on
+centralized parameter servers for coordination between workers. An alternative
+is a decentralized approach in which workers communicate with each other
+directly via the MPI `allreduce` primitive, without using parameter servers.
+NVIDIA's [NCCL](https://developer.nvidia.com/nccl) library has implemented most
+of the MPI primitives efficiently on GPUs, and Uber's
+[Horovod](https://github.com/uber/horovod) makes it simple to do multi-gpu and
+multi-node training using TensorFlow; the TensorFlow team is also working on an
+`allreduce`
+[implementation](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/contrib/distribute).
+Compared to parameter servers, this approach arguably is more
+[bandwidth optimal](http://www.cs.fsu.edu/~xyuan/paper/09jpdc.pdf) and
+[scales better](https://eng.uber.com/horovod/).
+
+However, right now there is no easy way to launch this type of training jobs on
+Kubernetes. By providing a CRD and a custom controller, we can make
+allreduce-style distributed training as simple as training on a single node.
+
+## Goals
+* Provide a common Custom Resource Definition (CRD) for defining a single-gpu,
+multi-gpu, or multi-node training job.
+* Implement a custom controller to manage the CRD, create dependent resources,
+and reconcile the desired states.
+* The cross-pod communication should be secure, without granting unnecessary
+permissions to any pod.
+* Though the initial version focuses on TensorFlow/Horovod, the approach can be
+applied to other frameworks with MPI support, such as CNTK.
```

Comment:
There is something that I am missing here.

It sounds like you could be doing multiple things here:

1) Allow using horovod in kubeflow (the simplest)
2) Allow running MPI jobs within kubeflow. In this case, how do we handle possible race conditions between both NCCL and MPI creating a ring for the computations?

Reply:
Horovod uses MPI and handles NCCL ring creation - so there will be no race condition.

Reply:
@alsrgv First of all, it is my understanding that you are the Horovod maintainer. Great to have you.

Maybe race condition is the wrong word. What I mean is that I am not sure what will happen if you use both CUDA-aware MPI and NCCL (i.e., you make a communication clique in both nccl and MPI) at the same time. (I was never really sure why you depend on both MPI and NCCL in Horovod, actually, but never took a deep enough dive into the code to understand why. Sumimasen). I mention this because this documentation sprinkles the word nccl around in a few places despite being primarily about MPI, so I wanted to point it out.

Reply:
Yes, I am - thanks :-)

Horovod uses NCCL because it's much faster than CUDA-aware MPI.  However, it still uses MPI for control communications between chief worker and everyone else - workers report which tensors they're ready to reduce and chief commands the order that is suitable for all.  This is necessary for TensorFlow since it has non-deterministic operation execution order in presence of forks in tue graph.

NCCL and MPI on CPU buffers will not deadlock.

Reply:
@everpeace added link to ChainerMN.